SRoC calculate charge controller

StuAA78 opened the pull request:
This change implements the calculate charge controller and its routing.

Several parts of the end-to-end process have been touched as part of this change -- putting all the pieces together helped me find several bits where the existing work didn't quite meet requirements:
- The Charge model validation had several things marked as required which aren't needed for SRoC charges;
- The Charge presenter didn't correctly return a positive or negative value based on `chargeCredit`;
- The Rules Service presenter didn't correctly return the financial year;
- The Calculate Charge service attempted to pull data from `response` instead of `response.body`;
- The Rules Service hadn't been changed to reflect the updated process (passing in a presenter instead of a translator);
- The Charge Translator was missing `chargeCategoryCode`;
- Validation has been moved from models to translators as it made more sense to put it there -- for example, this means that validation errors sent to the user return the parameter name sent to the API in the request (eg. `billableDays`) rather than the parameter name used in the model (eg. `regimeValue4`).

Cruikshanks reviewed (changes requested):
I think there are some small corrections to be made. Plus I have a suggestion about the fixtures.

Over than that it'd be great to be given a run-through as well.

StuAA78 commented:
As a note to myself -- I've just noticed that the charge translator doesn't seem to correctly convert dates; if the date is provided as `"01-APR-2020"` then it becomes `2020-03-31T23:00:00.000Z`. I'm not sure if it's somehow timezone-related but it's something that needs to be looked at.

StuAA78 commented on app/translators/base.translator.js:
```diff
@@ -1,20 +1,35 @@
     Object.defineProperty(this, '_data', { enumerable: false })
 
+    // Create each getter in _translations. We do this within the constructor to keep Joi happy
+    // TODO: We may not need to do this here now that validation is done on the presenter
```
Agh, I keep saying presenter instead of translator! The TODO is accurate with that in mind -- I originally placed the creation of getters within the constructor as creating them outside of it caused validation to fail when passing the presenter to the model. So in theory we could now move creation of the getters outside of the constructor now that validation is done by the translator. However I don't recall right now if that's even necessary/best practice to do so.

StuAA78 commented on test/controllers/calculate_charge.controller.test.js:
```diff
@@ -0,0 +1,120 @@
+  // Create auth token for stubbing authorisation
+
+  /**
+   * TODO: This only works with a non-admin token if the preceding Airbrake tests are skipped; I suspect something
```
Yes, this sorted it, thanks 👍

StuAA78 commented on test/controllers/calculate_charge.controller.test.js:
```diff
@@ -0,0 +1,120 @@
+  })
+
+  afterEach(async () => {
+    Nock.cleanAll()
```
Done 👍

StuAA78 commented on test/presenters/charge.presenter.test.js:
```diff
@@ -0,0 +1,36 @@
+// Test framework dependencies
+const Lab = require('@hapi/lab')
+const Code = require('@hapi/code')
+const Sinon = require('sinon')
```
Done 👍

StuAA78 commented on test/translators/base.translator.test.js:
```diff
@@ -5,30 +5,62 @@ const Lab = require('@hapi/lab')
+    })
 
-    const testTranslator = new BaseTranslator(testData)
+    it('throws an errror if data is valid', async () => {
```
🤦

StuAA78 commented on test/controllers/calculate_charge.controller.test.js:
```diff
@@ -0,0 +1,120 @@
+// Things we need to stub
+const JsonWebToken = require('jsonwebtoken')
+
+const rulesServiceRequestFixtures = [
```
👍

StuAA78 commented:
I've made the following changes based on post-review comments:
- Moved controller functions into class methods
- Have refactored translation getters so we now just create them as properties of the object
- Have added ability to specify a proxy for the rules service
- Have added an environment variable `SROC_MINIMUM_DATE` so we can set this to 2020 while waiting for the 2021 rules service endpoint to be created
- Added `TZ=UTC` to the environment variables to enforce UTC, fixing the charge translator date issue I spotted

Cruikshanks commented on .env.example:
```diff
@@ -21,6 +24,7 @@ PORT=3000
 RULES_SERVICE_URL='https://rules.service'
 RULES_SERVICE_USER='username'
 RULES_SERVICE_PASSWORD='password'
+RULES_SERVICE_PROXY='http://proxy.server:3128'
```
Had a discussion already. Just follow convention we're going to rely on an env var named `HTTP_PROXY` to be present if a proxy is needed.

Recording my comment here should anyone come asking why this PR isn't merged yet 😁

Cruikshanks reviewed (approved):
![lgtm](https://user-images.githubusercontent.com/1789650/98838438-2dcd5080-243c-11eb-94a9-fc7df7b4789a.gif)
